File: linked_list/707_designed_linked_list.py
from dataclasses import dataclass
from typing import Optional


@dataclass
class Node:
    value: int
    next: Optional["Node"] = None

@dataclass
class MyLinkedListClassic:
    size: int = 0
    head: Node | None = None
    
    def add_head(self, val: int):
        self.add_at_index(0, val)

    def add_tail(self, val: int):
        self.add_at_index(self.size, val)

    def get_by_index(self, index: int):
        if index < 0 or index >= self.size:
            return

        current = self.head        
        for _ in range(index):
            current = current.next
        
        return current.value

    
    def add_at_index(self, index: int, val: int):
        if index < 0 or index > self.size:
            return
        
        self.size += 1
        new_node = Node(val)

        if index == 0:
            new_node.next = self.head
            self.head = new_node
        else:
            current = self.head
            for _ in range(index - 1):
                current = current.next
            old_next = current.next
            current.next = new_node
            new_node.next = old_next
    
    def delete_by_index(self, index: int):
        ...
        

File: linked_list/707_designed_linked_list.go
type Node struct {
	Value int
	Next  *Node
}

type MyLinkedList struct {
	Capacity int
	Head     *Node
}

func Constructor() MyLinkedList {
	return MyLinkedList{}
}

func (this *MyLinkedList) Get(index int) int {
	if index < 0 || index >= this.Capacity {
		return -1
	}

	current := this.Head
	for i := 0; i < index; i++ {
		current = current.Next
	}

	return current.Value
}

func (this *MyLinkedList) AddAtHead(val int) {
	this.AddAtIndex(0, val)
}

func (this *MyLinkedList) AddAtTail(val int) {
	this.AddAtIndex(this.Capacity, val)
}

func (this *MyLinkedList) AddAtIndex(index int, val int) {
	if index < 0 || index > this.Capacity {
		return
	}
	this.Capacity++

	if index == 0 {
		newNode := &Node{val, this.Head}
		this.Head = newNode
		return
	}

	current := this.Head
	for i := 0; i < index-1; i++ {
		current = current.Next
	}
	oldNext := current.Next
	current.Next = &Node{val, oldNext}
}

func (this *MyLinkedList) DeleteAtIndex(index int) {
	if index < 0 || index >= this.Capacity {
		return
	}

	this.Capacity--

	if index == 0 {
		this.Head = this.Head.Next
		return
	}

	current := this.Head
	for i := 0; i < index-1; i++ {
		current = current.Next
	}

	current.Next = current.Next.Next
}
